### ISSUE TITLE
Is binary searching length of a file still necessary?

### ISSUE THREAD

**Issue by arter97** [landley/toybox#173]
I've noticed that fdlength() at lib/lib.c uses a very weird method to get the file size:

```
  // If the ioctl works for this, return it.
  // TODO: is blocksize still always 512, or do we stat for it?
  // unsigned int size;
  // if (ioctl(fd, BLKGETSIZE, &size) >= 0) return size*512L;

  // If not, do a binary search for the last location we can read.  (Some
  // block devices don't do BLKGETSIZE right.)  This should probably have
  // a CONFIG option...

  // If not, do a binary search for the last location we can read.
```

This breaks mkswap'ing a block device with broken read(2): e.g. a vnswap device.

Granted, vnswap is not a typical block device as it's meant to be exclusively used within the kernel, but it's still weird seeing such method still in use.

I don't know the rationale behind using a binary search as it predates the original code introduction in commit 055cfcbe5b0534c700b30216f54336b7581f7be4 , but SEEK_END would suffice imho(and it fixes vnswap).

**Comment by landley** [comment]
Sorry, I replied to this as mailing list email before realizing it was github. http://lists.landley.net/pipermail/toybox-landley.net/2020-March/011515.html

Try https://github.com/landley/toybox/commit/e8705902647ae62bdf3c7605d76be725ef385a63

**Comment by arter97** [comment]
> And to answer the question I left myself in the comment, by looking at the linux
> kernel source in block/ioctl.c:

I've just confirmed that BLKGETSIZE works fine even with 4K block devices.
Maybe it's hardcoded to 512 for compatibility...

> Breaks how? (You have a write-only device?)

Kind of. This is a dummy block device that's meant to be used with zswap made by Samsung.
Samsung shipped this until the Galaxy S9(ish), so I must assume they workarounded this issue on their own tree.

Write also fails at some point, but since it's just a dummy registered device, kernel doesn't have to care, but mkswap must succeed.

> It should be the fallback if the ioctl fails.

I've just tried out the new patch, but it doesn't work for vnswap.

ioctl() returns success but mkswap says:
`Swapspace size: 2147483644k, UUID=c789b9ea-fde6-41ca-9d9f-cb404c58aa19`

The correct size should be `4194300k`.

But this is the fault of vnswap not reporting block size correctly, not a fault of toybox, so I think we can close the issue here.

> In the case of a CDROM I think it's that the ioctl didn't spin up the disk, but
> the kernel didn't know how big it was yet if it wasn't spun up? (This was a
> while back.)

Is there a reason why SEEK_END would fail for a CDROM?
I think SEEK_END is much cleaner than the binary search.

Thanks.
(ps, I've sent out a dmesg patch too. Just a bump in case you missed it)
